feat: install multiple JDKs with same major version

quintesse opened the pull request:
Fixes #2118


<!--
Thanks for submitting your Pull Request!

Please delete this text, and add description of the topic solved by this PR.

To make releases as automated as possible we use conventional commits formats.
Thus commits and pull-requests titles should before merge follow the Conventional Commits spec.
See https://www.conventionalcommits.org/

Details in https://github.com/Ezard/semantic-prs

If in doubt then open the pull-request and we'll help you - Thank you!
-->

quintesse committed:
feat: install multiple JDKs with same major version

Fixes #2118

maxandersen commented:
Interesting. How's it to be used/managed? Couldn't spot clearly the user impact if any?

quintesse committed:
feat: allow setting `jdk default` per version

Added a flag `--for-version` (or `-v`) to the `jdk default` command to
allow setting the default Jdk to use for a specific major version.
This is necessary now that we allow multiple Jdks of the same major
version to be installed simultaneously.

quintesse commented:
@maxandersen this is now ready for review and devkitman 0.4.0 [has been released](https://github.com/jbangdev/jbang-devkitman/releases/tag/v0.4.0) so you don't need to build that to be able to test this PR.

Given the fact that this is a pretty major change it must be tested quite well (I already did of course, but I might be "blind" to certain issues). It must also be tested that older JBang versions keep working even when newer JDKs get installed by this new version.

Also, the workings and output of some of the jdk commands (eg `jdk list`) have changed somewhat, we might still want to polish that.

maxandersen reviewed (commented):
Should I ignore or honor the "do not merge" message ? :)

quintesse commented:
Ok, I marked this PR ready for review again because everything works now. The integration tests are failing because of the Alpine test not being able to download the Jdk 17 it needs, but that's unrelated to this PR.

quintesse committed:
feat: added `--jdk-installer` option

Added some more parameter checks to `jdk install`.
Also refactored `Jdk` class.

quintesse commented:
> .so this is showing all major java versions, if available, ordered by vendors?

Yes, something like that. It's following the same we do right now. You have to imagine that by default our setting for vendors is "temurin,aoj" (preference given to temurin for being first in the list). And in that case doing `list -a` does the same: it gives only a single option per major version. (We could change that of course, but then it would change for the default as well. So then you're left wondering: "I see 2 JDK 17s, which one will get installed when I do `install 17`?)

When you ask for details no filtering is performed.

quintesse committed:
feat: can now list providers and distros

quintesse committed:
chore: updated native image configuration

quintesse committed:
chore: fixed tests

maxandersen commented:
@quintesse any idea of why the new devkitman seem to fail handling idk 11 and other installs?

maxandersen commented:
seems to be wiremock related maybe?

maxandersen commented on src/main/java/dev/jbang/cli/JdkProvidersMixin.java:
```diff
@@ -9,16 +9,33 @@
 import picocli.CommandLine;
 
 public class JdkProvidersMixin {
+	protected static List<String> jdkProviders;
```
did you mean these to be static?  they depend on arguments set on command line so sure can be static, but what if/when run during tests?

quintesse replied:
But aren't the tests run sequentially?
Because if this is an issue we will have problems with all of these as well:

```
	private static boolean verbose;
	private static boolean quiet;
	private static boolean offline;
	private static boolean fresh;
	private static boolean ignoreTransitiveRepositories;
	private static boolean preview;
```

quintesse replied:
Ah yes, I remember now why they are static. It's because the options can be set on different "levels", if you type `jbang jdk --jdk-providers list` or `jbang jdk list --jdk-providers` the option gets handled by a different instance of the `JdkProvidersMixin` (and you could even mix! Imagine `jbang jdk --jdk-distros xxx list --jdk-providers yyy`, one option would be set in one instance and the other in a second instance). This makes figuring out what options to actually use much harder than it needs to be. So switching to static "fixes" that.

It's definitely not thread-safe, but the cli commands are not designed to be.

Our project compiler/builder API is much more thread-safe in this respect, but it still uses the static fields I mentioned above. That might be something we'd want to fix at some point.

maxandersen replied:
but shouldn't we then use picocli @parentcommand or scopetype = inherited?

maxandersen replied:
opened https://github.com/quintesse/jbang/pull/152 with pr that i *think* shows and fix the issue

quintesse commented:
Ok, all green, So please go ahead, play around with it and break things :-)

maxandersen commented:
i can't get any of the commands to work... doesn't recognize the jdk-providers/jdk-distros ...can't see tests use them either..what am I doing wrong?

quintesse commented:
> doesn't recognize the jdk-providers/jdk-distros

what do you mean "doesn't recognize"?

I can do the following without issue:

```bash
just jbang jdk l --jdk-providers=javahome
just jbang jdk --jdk-providers=javahome l
just jbang run --jdk-providers=javahome echo@jbangdev
```

And there are tests for `--jdk-providers`. Not for `--jdk-distros` indeed , but that's harder to test in the JBang project (because we'd need to actually download stuff and I wanted to avoid that). But in the devkitman project we can properly mock things and it's tested there. But we should indeed come up with _some_ way to test it with jbang itself.

maxandersen commented:
ok, it does work now. i got caught up in jdk-vendors vs jdk-distros. good stuff!

maxandersen commented:
wait - didnt actually test the multi isntall...

maxandersen commented:
I ran `just jbang jdk install --force --jdk-distros=graalvm 24` (it wanted --force exactly there - as had already a 24)
but it completed very fast (no download and no output) and I didnt get any linking?

```
ls -F ~/.jbang/cache/jdks
Permissions Size User Date Modified Name
drwxr-xr-x@    - max  13 Nov  2025   8/
drwxr-xr-x@    - max  23 Oct  2025   11/
drwxr-xr-x@    - max   9 Oct  2025   17/
drwxr-xr-x@    - max   9 Oct  2025   21/
drwxr-xr-x@    - max   7 Jan 10:35   22/
drwxr-xr-x@    - max  28 Dec  2025   24/
drwxr-xr-x@    - max  30 Sep  2025   25/
drwxr-xr-x@    - max  17 Mar 14:21   26/
```

bug?

quintesse commented:
Ok, will look into that.

But what would be the expected result here?

Because I could think of two different interpretations:

1. "Yes, I know I already have a 24, but I also want this one"
2. "Yes, I know I already have a default 24, but I also want this one and I want you to make it the default"

There could even be a 3rd interpretation:

3. "Yes, I know I already have a 24 but I want to use this one instead, so replace existing with this new one"

And depending on the other arguments you pass you might even expect a different result:

- `jbang jdk install --force 24` - Option (3)? Because there's a new minor release and you want to update to that one? Although option (2) might also work if we don't want to automatically remove versions
- `jbang jdk install --force --jdk-distros=graalvm 24` - Option (1)? Because it's obviously a different one? Or still option (3)? (Or option (2) if we don't want to auto-remove)

Option (1) might be difficult to explain, someone does `jbang jdk install --force --jdk-distros=graalvm 24` and then probably expects to be able to run `jbang --java 24 graalvmtest.java` and not first have to look up the id of the newly installed JDK and use that.

Looking at this option (2) is perhaps the one easiest to explain and giving the least headaches?

maxandersen commented:
I assume if i say use Zulu 24 that 24 will give me Zulu. 

I think that's what you call option 2. 

But this also mean we need someway to say "don't put this as default" 

Maybe Jbang jdk install --extra 24 ?

And somehow print what id to use if want to be explicit.

quintesse commented:
Not sure, that `--extra` isn't really an option that makes me "happy", so to speak.

It seems like a fix/workaround that's only useful in very specific situations that will be somewhat hard to explain when to use.

Perhaps combining option (1) with a `--default` option?

(THe problem might be: the default of what? There is a main/global default and a per-version default. DOes this option mean you want it to become the default for 24 or the main/global default?)

maxandersen commented:
How is the current PR intended to be used? it says install multiple jdks with same major version ...whats the command to do that if not `jbang jdk install --jdk-provider=zulu 24` ? how to choose which exact version will be installed?

quintesse commented:
Well the current PR doesn't yet give you all the options that I have in mind, because the changes were already big enough and I first wanted to get basic functionality working.

Anyway, if you do `jdk list -a -d` (`--available --details`) you'll see IDs and you can use those IDs to install _exactly_ that version, eg:

```
$ jbang jdk list -a -d
   27 (27-ea+17, jbang, 27-ea+17-temurin-jbang, [ea])
   26 (26+35, jbang, 26+35-temurin-jbang)
   25 (25.0.2+10, jbang, 25.0.2+10-temurin-jbang)
   ... etc
```

And then you can run `install 26+35-temurin-jbang` to install that exact version.

Right now we don't really show any other options, but you could easily imagine a `list` that returns options from multiple distros (like sdkman does, for example).

Now that we have the basics working we just have to come up with good ideas and what options/features to add :-)

quintesse commented:
For example, one of the things I want to allow is `<version>,<tag>...` where we now support `<version>`

Imagine:

- `install 24,zulu` instead of `install --jdk-provider=zulu 24`
- `install 24,jre,javafx` install a version 24 JRE with JavaFx support
- `run --java 24+,native` run using a 24 or higher GraalVM-based version


PD: I say `<version>`, but we actually support `<version>|<id>`, so the new format with tags would also be: `<version>,<tag>...|<id>`

maxandersen commented:
about provider and tags - thats why I was suggesting more structured format, like `24{jre,javafx}@zulu` which matches how we been discussing other ways to provide properties to jbang directive syntaxes.

maxandersen commented:
https://github.com/jbangdev/jbang/pull/2133 is where that syntax mostly implemented.

quintesse commented:
Sure, the actual syntax is irrelevant to devkitman, it's something that would be implemented by JBang itself and then implemented using the existing filtering system.

quintesse commented:
> I ran just jbang jdk install --force --jdk-distros=graalvm 24 (it wanted --force exactly there - as had already a 24)
but it completed very fast (no download and no output) and I didnt get any linking?

Btw, it seems the current jbang doesn't handle this either.
Still looking into it...

quintesse committed:
fix: force-installing a JDK should now work

quintesse commented:
> I ran just jbang jdk install --force --jdk-distros=graalvm 24 (it wanted --force exactly there - as had already a 24)
but it completed very fast (no download and no output) and I didnt get any linking?

@maxandersen this should be fixed now, can you try this new version?

maxandersen commented:
finally got around trying this again.

this time I seem to be getting a graalvm downloaded - yaay!

one thing I did spot is a repeated read of release...is that expected? 

```
just jbang --verbose jdk install --force --jdk-distros=graalvm 24
PATH="build/install/jbang/bin:$PATH" jbang --verbose jdk install --force --jdk-distros=graalvm 24
[jbang] [0:163] jbang version 0.138.0.7
[jbang] [2026-05-14 11:21:27] Using JDK provider(s): CurrentJdkProvider, DefaultJdkProvider, JavaHomeJdkProvider, PathJdkProvider, LinkedJdkProvider, JBangJdkProvider
[jbang] [2026-05-14 11:21:27] Unable to read 'release' file in path: /Users/max/.jbang/cache/jdks/8
[jbang] [2026-05-14 11:21:27] Unable to read 'release' file in path: /Users/max/.jbang/cache/jdks/8
[jbang] [2026-05-14 11:21:27] Unable to read 'release' file in path: /Users/max/.jbang/cache/jdks/8
[jbang] [2026-05-14 11:21:27] Unable to read 'release' file in path: /Users/max/.jbang/cache/jdks/8
[jbang] [2026-05-14 11:21:27] Unable to read 'release' file in path: /Users/max/.jbang/cache/jdks/8
[jbang] [2026-05-14 11:21:27] Unable to read 'release' file in path: /Users/max/.jbang/cache/jdks/8
[jbang] [0:348] Using cached file /Users/max/.jbang/cache/urls/1bee4c4e888dcfd526c7e4206b9921908844e651d2951c3565b28ac9b7e76c89/packages for remote https://api.foojay.io/disco/v3.0/packages?distro=graalvm&javafx_bundled=false&libc_type=libc&directly_downloadable=true&archive_type=tar.gz&operating_system=mac&package_type=jdk&version=24&release_status=ga&architecture=aarch64&latest=available
[jbang] [2026-05-14 11:21:27] Downloading JDK 24. Be patient, this can take several minutes...
[jbang] [2026-05-14 11:21:27] Downloading https://api.foojay.io/disco/v3.0/ids/07b0c04a8d6a8b0fa8911aec54b2e0e2/redirect
[jbang] [0:402] Requesting HTTP GET https://api.foojay.io/disco/v3.0/ids/07b0c04a8d6a8b0fa8911aec54b2e0e2/redirect
[jbang] [0:402] Headers {User-Agent=[JBang/0.138.0.7 (Mac OS X/26.3/aarch64) avaJ/25.0.2/Oracle Corporation]}
[jbang] [1:202] Redirected to: https://download.oracle.com/graalvm/24/archive/graalvm-jdk-24.0.2_macos-aarch64_bin.tar.gz
[jbang] [1:621] Deleting folder /Users/max/.jbang/cache/urls/e84da4bb18c108cdda486657256c89f2b7eeedfed1d921c53d698a53f4f65442.tmp
[jbang] [1:630] Deleting folder /Users/max/.jbang/cache/urls/e84da4bb18c108cdda486657256c89f2b7eeedfed1d921c53d698a53f4f65442-meta.tmp
^[[<65;137;39M^[[<65;137;39M^[[<64;127;44M[jbang] [16:980] Downloaded file https://download.oracle.com/graalvm/24/archive/graalvm-jdk-24.0.2_macos-aarch64_bin.tar.gz
[jbang] [2026-05-14 11:21:44] Installing JDK 24...
[jbang] [2026-05-14 11:21:44] Unpacking to /Users/max/.jbang/cache/jdks/24.0.2-graalvm-jbang
```

Whats the "Unable to read 'release' file in path: /Users/max/.jbang/cache/jdks/8" x 6 about?

maxandersen commented:
So after I ran `just jbang --verbose jdk install --force --jdk-distros=graalvm 24` I kinda expected `jdk list` to show some link to what I installed..

```
jbang jdk list --show-details
Installed JDKs (<=default):
   25 (25.0.2+10-69, current, current, /Users/max/.local/share/mise/installs/java/25.0.2)
   25 (25+36-LTS, default, default, /Users/max/.jbang/currentjdk) <
   25 (25.0.2+10-69, javahome, javahome, /Users/max/.local/share/mise/installs/java/25.0.2)
   25 (25.0.2+10-69, path, path, /Users/max/.local/share/mise/installs/java/25.0.2)
   24 (24.0.2+11-jvmci-b01, linked, 24, /Users/max/.jbang/cache/jdks/24)
   11 (11.0.28+6, jbang, 11-jbang, /Users/max/.jbang/cache/jdks/11)
   17 (17.0.16+8, jbang, 17-jbang, /Users/max/.jbang/cache/jdks/17)
   26 (26-beta+35-ea, jbang, 26-jbang, /Users/max/.jbang/cache/jdks/26)
   8 (1.8.0_472, jbang, 8-jbang, /Users/max/.jbang/cache/jdks/8)
   21 (21.0.8+9-LTS, jbang, 21-jbang, /Users/max/.jbang/cache/jdks/21)
   22 (22.0.2+9, jbang, 22-jbang, /Users/max/.jbang/cache/jdks/22)
   25 (25+36-LTS, jbang, 25-jbang, /Users/max/.jbang/cache/jdks/25)
 ```

the graalvm 24 is this:

`24 (24.0.2+11-jvmci-b01, linked, 24, /Users/max/.jbang/cache/jdks/24)`

but what is linked? and remind me what does 24 mean here vs 26-jbang?

quintesse commented:
Yay! Only took 7 months 😁
